jib-core/src/integration-test/java/com/google/cloud/tools/jib/api/ReproducibleImageTest.java

```diff
@@ -0,0 +1,194 @@
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.function.BiConsumer;
+import java.util.zip.GZIPInputStream;
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+/**
+ * Verify that created image has explicit directory structures, default timestamps, permissions, and
+ * file orderings.
+ */
+public class ReproducibleImageTest {
+  private static String LAYERS_PATH_IN_RESOURCES = "layers/";
+
+  @ClassRule public static final TemporaryFolder imageLocation = new TemporaryFolder();
+
+  private static File imageTar;
+
+  @BeforeClass
+  public static void createImage()
+      throws InvalidImageReferenceException, InterruptedException, CacheDirectoryCreationException,
+          IOException, RegistryException, ExecutionException {
+
+    Path root = imageLocation.getRoot().toPath();
+    Path fileA = Files.createFile(root.resolve("fileA.txt"));
+    Path fileB = Files.createFile(root.resolve("fileB.txt"));
+    Path fileC = Files.createFile(root.resolve("fileC.txt"));
+    Path subdir = Files.createDirectory(root.resolve("dir"));
+    Path fileD = Files.createFile(subdir.resolve("fileD.txt"));
+
+    imageTar = new File(imageLocation.getRoot(), "image.tar");
+    Containerizer containerizer =
+        Containerizer.to(TarImage.named("jib-core/reproducible").saveTo(imageTar.toPath()));
+
+    ExecutorService executorService = Executors.newCachedThreadPool();
+    containerizer.setExecutorService(executorService);
+    Jib.fromScratch()
+        .setEntrypoint("echo", "Hello World")
+        .addLayer(ImmutableList.of(fileA), AbsoluteUnixPath.get("/app"))
+        // layer with out-of-order files
+        .addLayer(ImmutableList.of(fileC, fileB), "/app")
+        .addLayer(
+            LayerConfiguration.builder()
+                .addEntryRecursive(subdir, AbsoluteUnixPath.get("/app"))
+                .build())
+        .containerize(containerizer);
+    Assert.assertFalse(executorService.isShutdown());
+    executorService.shutdown();
+  }
+
+  @Test
+  public void testImageLayout() throws IOException {
+    Set<String> paths = new HashSet<>();
+    layerEntriesDo(
+        (layerName, layerEntry) -> {
+          if (layerEntry.isFile()) {
+            paths.add(layerEntry.getName());
+          }
+        });
+    Assert.assertEquals(
+        ImmutableSet.of("app/fileA.txt", "app/fileB.txt", "app/fileC.txt", "app/fileD.txt"), paths);
+  }
+
+  @Test
+  public void testTimestampsEpochPlus1s() throws IOException {
+    layerEntriesDo(
+        (layerName, layerEntry) -> {
+          if (layerEntry.isFile() || layerEntry.isDirectory()) {
```

Comment:
nit: this test is unnecessary. In our simple example, every entry is either a file or a directory.

Reply:
That is true, but it worth checking this.  I'll pull the out to a separate test.